packages/sync-server/src/__tests__/CentralSyncManager.queueDeviceForSync.test.ts

```diff
@@ -0,0 +1,134 @@
+import { getTestModels } from '@tupaia/database';
+
+import { CentralSyncManager } from '../sync/CentralSyncManager';
+import { TestModelRegistry } from '../types';
+
+describe('CentralSyncManager.queueDeviceForSync', () => {
+  let centralSyncManager: CentralSyncManager;
+  let models: any;
+
+  const mockConfig = {
+    maxConcurrentSessions: 1,
+  };
+
+  beforeEach(async () => {
+    models = getTestModels() as TestModelRegistry;
+
+    centralSyncManager = new CentralSyncManager(models, mockConfig);
+
+    // Clear any existing data
+    await models.syncSession.delete({});
+    await models.syncQueuedDevice.delete({});
+  });
+
+  const requestSync = async (
+    deviceId: string,
+    lastSyncedTick: number = 0,
+    urgent: boolean = false,
+  ) => {
+    return centralSyncManager.queueDeviceForSync(`queue-${deviceId}`, urgent, lastSyncedTick);
+  };
+
+  const closeActiveSyncSessions = async () => {
+    await models.syncSession.update({}, { completed_at: new Date() });
+  };
+
+  it('Should start a sync if the queue is empty', async () => {
+    const result = await requestSync('A');
+    expect(result).toHaveProperty('sessionId');
+  });
+
+  it('Should queue if a sync is running', async () => {
+    const resultA = await requestSync('A');
+    expect(resultA).toHaveProperty('sessionId');
+
+    const resultB = await requestSync('B');
+    // we're first in line (device A having been removed from the queue to start its sync)
+    // but we are waiting for that sync to complete
+    expect(resultB).toHaveProperty('status', 'activeSync');
+
+    const resultC = await requestSync('C');
+    // we're behind device B
+    expect(resultC).toHaveProperty('status', 'waitingInQueue');
+  });
+
+  it('Should pick the oldest syncedTick device given uniform urgency', async () => {
+    const resultA = await requestSync('A'); // start active sync
+    expect(resultA).toHaveProperty('sessionId');
+
+    // get some sessions in the queue
+    await requestSync('B', 100);
+    await requestSync('C', 200);
+    await requestSync('D', 300);
+    await requestSync('E', 10);
+
+    await closeActiveSyncSessions();
+
+    const waiting = await requestSync('D', 300);
+    expect(waiting).toHaveProperty('status', 'waitingInQueue');
+
+    const started = await requestSync('E', 10);
+    expect(started).toHaveProperty('sessionId');
+  });
+
+  it('Should prioritise urgent over lastSyncedTick', async () => {
+    const resultA = await requestSync('A'); // start active sync
+    expect(resultA).toHaveProperty('sessionId');
+
+    // get some sessions in the queue
+    await requestSync('B', 100);
+    await requestSync('C', 200, true);
+    await requestSync('D', 300);
+    await requestSync('E', 10);
+
+    await closeActiveSyncSessions();
+
+    const waiting = await requestSync('E', 10);
+    expect(waiting).toHaveProperty('status', 'waitingInQueue');
+
+    const started = await requestSync('C', 200); // previous urgent flag should stick
```

Comment:
I don't think it actually tests that. When the `requestSync` for `E` happens, that's when the queue moves on after closing `A`, at which point there hasn't been any new `C` call without the urgent flag so it is chosen.

Please do add a test case where you do a `requestSync('C', 200)` _before_ the `closeActiveSyncSessions`, which will test the non-overwrite functionality.